developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,201 @@
+implemented, ideally in as idiomatic a fashion as possible.
+
+## Primitive Types
+
+The core primitives of the Pulumi type system form a superset of the JSON type system, and supports the
+following types:
+
+- `Null`, which represents the lack of a value
+- `Bool`, which represents a boolean value
+- `Number`, which represents an IEEE-754 double-precision number
+- `String`, which represents a sequence of UTF-8 encoded unicode code points
+- [`Asset`](#assets-and-archives), which represents a blob
+- [`Archive`](#assets-and-archives), which represents a map from strings to `Asset`s or
+  `Archive`s
+- [`ResourceReference`](#resource-references), which represents a reference to a resource
+- `Array<T>`, which represents a numbered sequence of values of a particular type
+- `Map<T>`, which represents an unordered mapping from strings to values of a particular
+  type
+- `Union<T0, T1 ... TN>`, which represents a value of one of a fixed set of types
+- `Enum<T, V0 ... VN>`, which represents one of a fixed set of values of a particular type
+
+### Assets and Archives
+
+An `Asset` or `Archive` may contain either literal data or a reference to a file or URL.
+In the former case, the literal data is a textual string or a map from strings to `Asset`s
+or `Archive`s, respectively. In the latter case, the referenced file or URL is an opaque
+blob or a TAR, gzipped TAR, or ZIP archive, respectively.
+
+Each `Asset` or `Archive` also carries the SHA-256 hash of its contents. This hash can be
+used to uniquely identify the asset (e.g. for locally caching `Asset` or `Archive`
+contents).
+
+### Resource References
+
+A `ResourceReference` represents a reference to a resource . Although
+all that is necessary to uniquely identify a resource within the context of a stack is its
+URN, a `ResourceReference` also carries the resource's ID (if it is not a component)) and
+the version of the provider that manages the resource. If the contents of the referenced
+resource must be inspected, the reference must be resolved by invoking the `getResource`
+function of the engine's builtin provider. Note that this is only possible if there is a 
+connection to the engine's resource monitor, e.g. within the scope of a call to `Construct`.
+This implies that resource references may not be resolved within calls to other 
+provider methods. Therefore, configuration values, custom resources and provider functions
+should not rely on the ability to resolve resource references, and should instead treat
+resource references  as either their ID (if present) or URN. If the ID is present and
+empty, it should be treated as an [`Unknown`](#unknowns).
+
+## Object Types
+
+Object types are defined as mapping from property names to property types. Duplicate
+property names are not allowed, and each property name maps to a single type.
+
+## `Promise<T>`
+
+Like promises in other systems, a value of type `Promise<T>` represents the result of an
```

Comment:
I think starting with a  reference to "other systems" is more confusing then helpful. I think it's much clearer to say 
```
`Promise<T>` represents the result of an asynchronous computation of type `T` or it's failure. 
```
As an aside, what is the type of a failure to resolve a promise? Should we mention that type (if it exists).

Reply:
> I think starting with a reference to "other systems" is more confusing then helpful.

Fair enough--I'll make that change 👍 

> As an aside, what is the type of a failure to resolve a promise? Should we mention that type (if it exists).

There isn't one. This is kind of a key difference from promises in other systems: it's not possible to handle the rejection of an output. This isn't really intentional, and we may want to reconsider it.

Reply:
I think that `represents the result of an asynchronous computation or it's failure` (I parse it `represents the result of ((an asynchronous computation) | (it's failure))` implies that `Promise<T>` is really `Promise<T, E> where E: Error`.

In light of your response, I would instead say that 
```
A value of type `Promise<T>` represents the result of a asynchronous computation.
```
 While a promise may of course fail, it doesn't seem like that failure is part of (can be represented in) the Pulumi type system.

Reply:
I had the same thought :)

> A value of type `Promise<T>` represents the result of an asynchronous computation. Note
> that although computations may fail, failures cannot be handled at runtime, and cause a
> hard stop when attempting to access a `Promise<T>`'s concrete value.